add check validity of edm::Handle in HLTDQM modules [forwardport of #24445]

mtosi opened the pull request:
forwardport of https://github.com/cms-sw/cmssw/pull/24445

mtosi committed:
add check validity of Handles and LogWarning

Update BPHMonitor.cc

Update HTMonitor.cc

Update METMonitor.cc

fix typo

mtosi commented:
@cmsbuild, please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24449/6281

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @mtosi (mia tosi) for master.

It involves the following packages:

DQMOffline/Trigger


@kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please review it and eventually sign? Thanks.
@battibass, @jhgoh, @calderona, @HuguesBrun, @trocino, @folguera, @rociovilar this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1
Tested at: 3538744fd64c5142bf16aca985e471b239bbeb66
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30229/summary.html

cmsbuild commented:
Comparison job queued.

fabiocos commented:
@jfernan2 @andrius-k @schneiml please check and sign it in case. The real solution looks to me adding the appropriate sequence to have the needed collection in the DQM sequence

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30229/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3143879
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3143681
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

fabiocos commented on DQMOffline/Trigger/plugins/BPHMonitor.cc:
```diff
@@ -364,13 +374,40 @@ void BPHMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
+    }
+    return;
+  }
+  if ( !trHandle.isValid() ) {
```
@mtosi why a second time? Is it a double cut-and-paste? I am missing the logic here...

mtosi replied:
:(

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24449/6288

cmsbuild commented:
Pull request #24449 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

fabiocos commented:
please test

davidlange6 commented on DQMOffline/Trigger/plugins/HTMonitor.cc:
```diff
@@ -201,36 +229,57 @@ void HTMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup)
+    if ( electrons.size() < nelectrons_ ) return;
+  } else {
+    if (!warningPrinted4token_[2]) {
+      edm::LogWarning("HTMonitor") << "skipping events because the collection " << jetInputTag_.label().c_str() << " is not available";
```
hi-
I generally find these changes distasteful - however one practical comment - there is nothing here that actually skips the event. instead any checks on the number of good electrons are just skipped Probably that is not the behavior desired. 
(note also that the variable electrons is completely unnecessary it seems)

fabiocos replied:
@davidlange6 as discussed directly we agree, I consider this as a quick workaround, the final solution is to properly add the producer in the correct sequence

mtosi replied:
if the input electron collection is wrong, because not available, I doubt one can rely on any outcome of this code
=> cutting or not on the electrons multiplicity loses its value, I think
said that, I'm almost sure not all the consumers / users of this module need electrons,
=> I preferred to keep a conservative approach, which allows to process the event

fabiocos replied:
@mtosi but you would need a return statement after setting the flag to true, right?

mtosi replied:
the selections on electron will not be done, but the plots will be (wrongly) filled ...
but, if one does not want to apply any cuts on the electrons multiplicity and indeed even misconfigured the module, the plot will be (properly) filled and fine, and the warning message will make people know to update the config

am I wrong ?

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24449/6290

cmsbuild commented:
Pull request #24449 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

mtosi commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5c5ae98323ad2e1549be5de47b6be5cff7b5ad2d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30241/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30241/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3143879
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3143681
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

mtosi committed:
moved softdrop modules to WithRECO sequence

fix logic

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24449/6295

cmsbuild commented:
Pull request #24449 was updated. @cmsbuild, @andrius-k, @Dr15Jones, @kmaeshima, @cvuosalo, @schneiml, @ianna, @mdhildreth, @nsmith-, @rekovic, @jfernan2, @thomreis, @civanch can you please check and sign again.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fabiocos commented on DQMOffline/Trigger/plugins/BPHMonitor.cc:
```diff
@@ -658,25 +688,31 @@ void BPHMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
+	  } else {
+	    if (!warningPrinted4token_[3]) {
+	      edm::LogWarning("BPHMonitor") << "skipping events because the collection " << phInputTag_.label().c_str() << " is not available";
+	      warningPrinted4token_[3] = true;
```
@mtosi don't you need a return statement here?

fabiocos replied:
sure, the code could still run, but it will give different results wrt the fully functional version. So depending on which input is missing you could get different results from the same code without immediately realising it (in the DQM checks, people will not normally look at log files).
It is a matter of choices, having no filled histogram is in my opinion perhaps preferable than having filled histograms filled in an unexpected way. But I think this is ultimately something that should be agreed by the end users. We are always speaking of pathological situations.

fabiocos commented on DQMOffline/Trigger/plugins/HTMonitor.cc:
```diff
@@ -201,36 +229,57 @@ void HTMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup)
+  } else {
+    if (!warningPrinted4token_[3]) {
+      edm::LogWarning("HTMonitor") << "skipping events because the collection " << vtxInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[3] = true;
```
@mtosi and a return statement here?

fabiocos commented on DQMOffline/Trigger/plugins/HTMonitor.cc:
```diff
@@ -201,36 +229,57 @@ void HTMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup)
+    if (!warningPrinted4token_[4]) {
+      edm::LogWarning("HTMonitor") << "skipping events because the collection " << muoInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[4] = true;
+    }
```
@mtosi and here...

fabiocos commented on DQMOffline/Trigger/plugins/METMonitor.cc:
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
+  } else {
+    if (!warningPrinted4token_[1]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << jetInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[1] = true;
```
@mtosi and here

fabiocos commented on DQMOffline/Trigger/plugins/METMonitor.cc:
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
+  } else {
+    if (!warningPrinted4token_[2]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << jetInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[2] = true;
```
@mtosi and here

fabiocos commented on DQMOffline/Trigger/plugins/METMonitor.cc:
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
+  } else {
+    if (!warningPrinted4token_[3]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << vtxInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[3] = true;
```
@mtosi and here

fabiocos commented on DQMOffline/Trigger/plugins/METMonitor.cc:
```diff
@@ -148,56 +159,93 @@ void METMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
+  } else {
+    if (!warningPrinted4token_[4]) {
+      edm::LogWarning("METMonitor") << "skipping events because the collection " << muoInputTag_.label().c_str() << " is not available";
+      warningPrinted4token_[4] = true;
```
@mtosi and here

fabiocos commented on L1Trigger/Configuration/python/SimL1Emulator_cff.py:
```diff
@@ -62,7 +62,8 @@
 
 from Configuration.Eras.Modifier_phase2_hgcal_cff import phase2_hgcal
-phase2_hgcal.toReplaceWith( SimL1Emulator , _phase2_siml1emulator )
+from Configuration.Eras.Modifier_phase2_hgcalV9_cff import phase2_hgcalV9
```
this change looks unrelated to the rest of the PR, it is just #24418 already merged picked up

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7837661e5243aeb9cfa61a2aeaaee2715d7d7a23
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30268/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7962133bd326ba03e4b13594892bce4162d6f416
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30268/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30268/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30268/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3143975
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3143777
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 6.536 KiB( 31 files compared)
 - DQMHistoSizes: changed ( 150.0,... ): 3.268 KiB        HLT/B2G
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

fabiocos commented:
@mtosi when you update the code, could you please rebase it onto a recent IB so as to avoid spurious insertion of unrelated commits?

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24449/6323

cmsbuild commented:
Pull request #24449 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

mtosi commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 125c5f3dd6d3228729348f5922e333938a387695
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24449/30275/summary.html

cmsbuild commented:
Comparison job queued.

mtosi commented:
@fabiocos could this PR be integrated and a patch release w/ this be built ?
TSG asked to PdmV to produce samples for the 102X validation w/ data, and we need this PR in

please, let me know if there is anything to do
and thanks !

fabiocos commented:
@jfernan2 @andrius-k @schneiml could you please check and sing it in case? We need to integrate the backport in 10_2_X to fix the present problem in the release

andrius-k commented:
Hi Mia,

It seems that changes in this PR and the original [backport](https://github.com/cms-sw/cmssw/pull/24445) don't match. Maybe you forgot to push one more commit here?

fabiocos commented:
@andrius-k I let @mtosi comment better, but as far as I can see the modified code is already different on input, and it looks to me that the differences in changes adapt to this original difference. Do you see places where this is not true?

mtosi commented:
HTMonitor.cc and METMonitor.cc are identical
while BPHMonitor is different in the 2 releases
=> I modified manually the 2 files
I introduced `vtxInputTag_` in the master, because there is the corresponding Token only here (even if it is not used by the code)
then, warningPrinted4token_ has a larger size in the master than 102X (see above comment)

the rest reflects the original differences between the master and 102X

@andrius-k are you referring to a special piece of code ?

thanks !

andrius-k commented:
+1
I just noticed that the number of commits and code changes are significantly different between the two PRs and was just double checking that this is intended.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
